udf_js/weighted_quantile.sql

```diff
@@ -0,0 +1,85 @@
+CREATE OR REPLACE FUNCTION udf_js.weighted_quantile(
+  percentile FLOAT64,
+  histogram ARRAY<STRUCT<key STRING, value FLOAT64>>,
+  type STRING
+)
+RETURNS FLOAT64
+LANGUAGE js
+AS
+  '''
+  if (percentile < 0 || percentile > 100) {
+      throw "percentile must be a value between 0 and 100";
+  }
+
+  let keys = histogram.map(bucket => parseInt(bucket.key));
+  let values = histogram.map(bucket => bucket.value);
+
+  return weightedQuantile([percentile/100], keys, values)[0];
+'''
+OPTIONS
+  (library = "gs://moz-fx-data-circleci-tests-bigquery-etl/wtdstats.js");
+
+SELECT
+  -- centered
+  assert_equals(
+    1,
+    udf_js.weighted_quantile(
+      50.0,
+      ARRAY<STRUCT<key STRING, value FLOAT64>>[("0", 10), ("1", 10), ("2", 10)],
+      "timing_distribution"
+    )
+  ),
+  -- skew right
+  assert_equals(
+    2,
+    udf_js.weighted_quantile(
+      50.0,
+      ARRAY<STRUCT<key STRING, value FLOAT64>>[("0", 10), ("1", 10), ("2", 100)],
+      "timing_distribution"
+    )
+  ),
+  -- skew left
+  assert_equals(
+    0,
+    udf_js.weighted_quantile(
+      50.0,
+      ARRAY<STRUCT<key STRING, value FLOAT64>>[("0", 100), ("1", 10), ("2", 10)],
+      "timing_distribution"
+    )
+  ),
+  -- skew right
+  assert_equals(
+    2,
+    udf_js.weighted_quantile(
+      50.0,
+      ARRAY<STRUCT<key STRING, value FLOAT64>>[("0", 10), ("1", 10), ("2", 100)],
+      "timing_distribution"
+    )
+  ),
+    -- Bug when all weights are less than 2. This test fails with a `400
```

Comment:
Yes, it seems to be that there are bugs from the few tests that I've added here. The idea is that the stats code would be factored out of GLAM so it can be reused here too, which would fix some of the bugs here. The UDF wouldn't be located within the repository, but rather an external library that could be referenced directly in gcs or uploaded to the udf lib bucket with some minor modifications.